### ISSUE TITLE
Current issues/bugs to resolve

### ISSUE THREAD

**Issue by birdybro** [MiSTer-devel/Arcade-TropicalAngel_MiSTer#1]
- [x] Initial boot of the rom seems incorrect, might need to slow down rom loading upon loading the core.
- [x] Occasionally the above mentioned rom loading can make the screen lose sync.
- [x] It's possible the SDRAM is not being flushed correctly, this might also be related to the above.
- [x] Ocean is purple, testing some r/b bit-order swapping to fix this now.
- [x] Someone reported input mapping issues? I didn't notice anything wrong with my syntax and it works for me, but I need to investigate further.
- [x] Need to change back to the original clocks of the core. I ran into significant timing issues previously, but now that I figured out other stuff it might not happen anymore, or I might find easy ways around it.
~~- [ ] Extend number of lines from 239 to 240 possibly to resolve csync mismatch to real hardware?~~ Irrelevant sorta
- [x] The sound effect for the Angel falling down is playing slightly too fast it seems. - switched to a clock with a /4 divider to simulate how it was done on the original board to rule out mistakes in my previous method, timing req is still met, so maybe it's better this way? Need to test. This maybe will fix it? --> https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/commit/6352ffafe4eb4a9706fe8b44b2b884427be4ee58
- [x] Sometimes on reset of the core the sound chip gets bugged out, not sure if this happens on MiST or not... Sound is still somewhat unstable... The reset needs to be handled more carefully I think.
- [ ] Bottom of the screen is missing, gyurco outlined the issues below, probably requires a reverse engineering of a proprietary IC that isn't RE'd yet.
- [x] Changing DIP options also resets the game which probably shouldn't happen, that's not like other cores, or maybe og hardware did work that way?

**Comment by birdybro** [comment]
Resolved the booting of the rom and screen sync that happens a result already, that was easy, with this --> https://github.com/birdybro/Arcade-TropicalAngel_MiSTer/commit/6f984097849fec32e1e5eef36e70fb1f06f4299b

Fixed the red bit-swap needed to color the ocean correctly here --> https://github.com/birdybro/Arcade-TropicalAngel_MiSTer/commit/e267657d98c2c32ef056ebb96caf59a7efb25bff

**Comment by kold669** [comment]
I own the board and to me it seems the sea colour is correct

**Comment by birdybro** [comment]
Thank you! The color is now similar to a picture from an auction that was shared with me thanks to @wwark 

![image](https://user-images.githubusercontent.com/16388068/201534528-79c425c7-395a-4790-955c-dec27f575651.png)

![image](https://user-images.githubusercontent.com/16388068/201534535-d696867c-d7d5-4b80-8969-54825931a1ef.png)

The bits must be inverted, and I did fix it with the latest release. Old MAME pictures show a purple ocean so the bits didn't used to be inverted.

I have a few questions if you have time. Since you have the board, do you have schematics for it or the arcade manual as well? I haven't been able to find those. I'm attempting to tackle the DIP switches, and I need confirmation that the information on arcade-italia is correct or not (http://adb.arcadeitalia.net/dettaglio_mame.php?game_name=troangel - scroll to the bottom and expand "show DIP switches" to see).

This is a board that has very scarce information on the internet and it would be awesome to confirm some things. I don't even know what the oscillators officially should be, I'm working off of limited information :D

Thank you for any and all help you are able to provide as always @kold669 ! :)

**Comment by kold669** [comment]
Here is a picture of my pcb to confirm the colours:

https://ibb.co/wsCWhLZ

Here is the dipswitch table I found at home ( from an arcade operator paperwork I got many years ago)

https://ibb.co/j8gGmtJ

Unfortunately I don't have any schematics of this board, but I think the hw is similar to Moon Patrol in someway

Main OSC is 18.432 ( Z80 runs at 3.072mhz)

Sound OSC is 3.57mhz 

Csync is 15.44khz and Vsync is 56.75hz

Hope these info are useful

**Comment by birdybro** [comment]
Thank you very much! That is very useful! :)

**Comment by kold669** [comment]
I measured the actual sound section clocks

6803E runs at 894khz like the AY8910 
MSM5205 runs at 384khz

Csync is 15.609khz ( not 15.44khz)

Mame has the 6803E running at 3.57mhz??

**Comment by birdybro** [comment]
Yes, the clock is inaccurate currently. I'm fixing that and you are helping me a lot with it! :)

Main oscillator of 18.432MHz makes sense, the core is doing a doubling of that, and then dividing down, at what will be 36.863998MHz when I get done testing it.

Sound oscillator of 3.57MHz makes sense, the core is using 0.895MHz, so it must have a /4 divider somewhere on the real board, or be doing /4 division at some point.

For the MSM5205 I don't have a separate clock to it in the core currently. Would be interesting to test that later, so I'll keep that in mind.

**Comment by birdybro** [comment]
@kold669 are the measurements on the MiSTer correct with this build:

[TropicalAngel_20221113.zip](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/files/9998386/TropicalAngel_20221113.zip)

**Comment by kold669** [comment]
I will check after work!

yes the 3.57 osc is 3.579545mhz to be precise and the sound cpu and ay get it divided by 4

the msm5205 has its own small quartz at 384khz

**Comment by birdybro** [comment]
I see the clock enable for 384KHz already, so it is using that correctly.

https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/blob/master/rtl/moon_patrol_sound_board.vhd#L220-L235

I need to double check the math on the clk_aud I'm using though, I think it might be slightly under clocked. Could be causing an underrun for the audio.

**Comment by kold669** [comment]
Tropical Angel is Irem M57 but the sound board is exactly the same as Moon patrol and Zippy Race  ( M52 Sound C).
With correct clock maybe it can fix the sound of the girl falling down that at first glance on the pcb has a lower pitch ( unless it's generated by thge 5205, not sure)

**Comment by birdybro** [comment]
~~When you put the DIPS into service mode are you able to do anything with the controls? I couldn't figure out how I was supposed to interact with the service mode. I want to see if I can isolate that sound effect.~~ Nevermind.

Actually I think I know what it might be if the MiSTer is running that sound effect slightly too fast. The clk_aud counter is set to the previous clk_sys speed as a reference, so I can just adjust that and it should fix it. Try this build for me please and compare and let me know if it sounds correct to you. :)

[TropicalAngel.zip](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/files/10004823/TropicalAngel.zip)

**Comment by kold669** [comment]
Last build stay stucked on the RAM ROM OK

**Comment by birdybro** [comment]
Make sure to update your MRA files to the latest and do "Reset Settings". One of the DIPS can cause that to happen.

**Comment by kold669** [comment]
From a first glance it seems matching with the pcb. Later I will check the board and let you know

**Comment by birdybro** [comment]
Thank you! :)

This is what possibly resolved it for reference:

![image](https://user-images.githubusercontent.com/16388068/201734509-f6ed0dd5-5903-4d91-a405-2b2e333a5de0.png)

**Comment by kold669** [comment]
I made a side by side comparison, the music and boat sound are correct.
The sound when you fall down it still off.

I recorded a small wav with mobile phone when you fall down, I think it's still clear:

[wav.zip](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/files/10006661/wav.zip)

Also the screen ratio seems off.  With the same settings of my monitor the core video appear streched vertically. 
Csync should be 15.6khz instead of 16khz.

**Comment by birdybro** [comment]
Yeah, I think the original core has 239 lines displayed instead of 240 for some reason which would be causing that. This has been the case in many MiST core ports and has to be adjusted afterward.

**Comment by kold669** [comment]
I think once the sound and screen ratio is addressed the core is pretty much perfect.
I still don t understand how the analogue gas can work since on the board there isn’t any analogue/digital ic

**Comment by birdybro** [comment]
According to MAME it might overlap with 4 of the dip switches, multiplexed or something. I'm tempted to just hook it up to an analog joystick and test it to see what it does. :D

**Comment by birdybro** [comment]
@kold669 care to test this one for the csync? I *think* I understood, but I haven't tested it yet personally. Just am curious to see if my guess is correct ;)

[TropicalAngel.zip](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/files/10007282/TropicalAngel.zip)

**Comment by kold669** [comment]
Csync is still at 16khz?

You are right, the bottom of the screen if missing some lines , probably it's the reason why the screen is stretched as you said

**Comment by kold669** [comment]
Core:
https://ibb.co/vz2rybP

PCB:
https://ibb.co/kyShK9m

Look at the bottom of the screen for the differences. ( colours are different because my phone has troubles with crt refresh)

**Comment by birdybro** [comment]
@kold669 

Can you try this build for me please? (I'm very new to this, thank you for your patience)

[TropicalAngel_20221121.zip](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/files/10059280/TropicalAngel_20221121.zip)

Thanks! :)

**Comment by kold669** [comment]
hi!
aspect ratio is good but there are some gfx glitches on the right side of the screen and the gfx on the bottom of the screen seems to be cut. If you add a credit, you can see the credit text being cut
Also I cannot hear anyore the sound of falling down

**Comment by birdybro** [comment]
Thank you for testing. I'll have to investigate that since it mirrors what I saw on my digital display as well. There appears to also be some instability added by this, probably a lot of extra logic is contingent upon these frequencies, so this could be a very long project :P

**Comment by kold669** [comment]
Let me know if i can help further 👍👍

**Comment by birdybro** [comment]
@gyurco - Corrado here has measurements you may be interested in. Instead of:

```
-- hcnt [x080..x0FF-x100..x1FF] => 128+256 = 384 pixels,  384/6.1439Mhz => 1 line is 62.5us (16.000KHz) 
-- vcnt [x0E6..x0FF-x100..x1FF] => 26+256 = 282 lines, 1 frame is 260 x 62.5us = 17.625ms (56.74Hz)     
```

I think it should be:

```
-- hcnt [x076..x0FF-x100..x1FF] => 138+256 = 394 pixels, 394/6.1439Mhz => 1 line is 64.1286us (15.5936KHz actual and 15.609KHz measured from pcb)
-- vcnt [x0ED..x0FF-x100..x1FF] =>  19+256 = 275 lines, 1 frame is 275 x 64.1286us = 17.635ms (56.71Hz actual and 56.75Hz measured from pcb)     
```

but I'm still very new to messing with video sync signals and blanks. I'm confused on the blanks as the image now looks like this:

https://user-images.githubusercontent.com/16388068/203199367-8f1a30aa-bfb9-4b66-a6fb-bc3a0ddfe8f7.mp4

So I think I need to adjust the blanks, but the formulas are kind of interesting for them, whenever I adjust the horizontal blanks both the horizontal and vertical resolution increase. Any advice?

Thank you.

**Comment by gyurco** [comment]
My only reference is the Moon Patrol board schematics, which is definitely 384 pixels/line (0x80-0x1ff), and 282 lines/frame (0xe6-0x1ff).
If you can identify the H and V buses on the board, then probably it can be measured with a logic analyser. I don't know if they're implemented with discrete logic or some custom chips (like KNA6021901). If it's discrete logic with 74LS161 counters, then it would be interesting to see how the preload pins(3-4-5-6) are connected (either high or low).

**Comment by kold669** [comment]
here you can find hi res scans of the pcb ( not mine)
http://www.ym2149.com/arcadepcb/pcb/irem.html

**Comment by gyurco** [comment]
The Hcounter bits 0-7 are connected to the custom chip kna6032701 pins 13-20, which seems are originating from one of the connectors, so they're generated somewhere on the other board, with discrete logic. Good luck to reverse engineer it, it's not impossible, just very tedious. Or just simply connect these pins to a logic analyzer and check the progressing (should go to 0x00-0xFF then 0x80-0xFF).
As this chip is also used on the M62, I assume the timings should be the same as Kung-Fu Master, where two KNA6021901s do the H/V counter, they were replicated, and probably some info about its timings can be found somewhere.

**Comment by birdybro** [comment]
@kold669 Has the sound improved in the latest release I just pushed out?

Also please try the two color palette options and let me know what you think compared to your board.

**Comment by kold669** [comment]
Seems you corrected the sound yes.
Original option seems to match better the pcb

**Comment by birdybro** [comment]
Then it's somewhat confirmed, the sound was the result of the core not meeting timing closure I suppose. Thank you so much for your help @kold669!!

@gyurco You might be interested, check the [red bit order changes](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/blob/master/Arcade-TropicalAngel.sv#L384-L392) and the OSD option swap added.

Original:

![Original](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/assets/16388068/d5c0e2c1-b846-441e-b29f-3892cedac237)

Alternate:

![Alternate](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/assets/16388068/e43ba7ea-0255-4ac8-a3d5-d37759197dd0)

Personally it looks better and it matches the look of this crt picture that has calibrated output:

![calibrated_crt_arcade](https://github.com/MiSTer-devel/Arcade-TropicalAngel_MiSTer/assets/16388068/6b6b9deb-0e91-4bd5-ba1b-f8ed8bdeab82)
